DOCS-7780 4.3 DW Update-syntax-kt (#1321)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dw-operators.adoc

include::_attributes.adoc[]
endif::[]

> DataWeave 2.0 supports several mathematical, equality, relational, logical, prepend, append, flow control, scope, and update operators.

Reviewer: This list bothers me a little because "prepend" and "append" are verbs rather than adjectives: prepend operator? append operator? 
It would bother me less if they were not in a list with obvious adjectives.
So...are these TYPES of operators or the NAMES of operators? 
Maybe we can avoid this minor awkwardness by using names or syntax?
Or...we could use simply "DataWeave 2.0 supports several operators.", "...  types of operators."

Reviewer: AFTER EDITING FURTHER...can we present the operators in backticks? I don't know if that will work for the first four, however. Aaaaargh!

Reviewer: List is missing "remove."
We could say something that doesn't require us to name them all: 
"DataWeave 2.0 supports several operators, including mathematical operators, equality operators, and operators such as `prepend` and `append`."

| `/`  | For division.
|===

> The following example use mathematical operators:

Reviewer 2: `uses`?

> The following example use mathematical operators:

Reviewer 2: `uses`? same below

> The following example uses mathematical operators:

Reviewer: I wonder if "illustrates" would be better than "uses"? GLOBAL, if you think so.

the `!` only applies to the first `true`. `!(true or true)` returns
`false`.

> The following examples show uses of logical operators:

Reviewer: Yet another variation: "examples show uses of" (vs. "uses" or "illustrates")
If you can, standardize the wording of the intro. texts throughout this topic set, for types of operators.

(without the space) does not work.

> You can use logical operators together. The following example uses `or not` as defined in the `orNot` expression, uses `and not` in `andNot`, and uses `not` and `and not` in `notWithAndNot`.

Reviewer: "You can use logical operators together." > "You can use more than one logical operator in a <whatever>."?

Reviewer: End with a colon to point to the example.

Reviewer: An alternative is to make this narrative list into a bullet list:
```suggestion
You can use logical operators together. The following example uses <four> operators for <one ???>:
* `or not` as defined in the `orNot` expression
* `and not` in `andNot`
* `not` and `and not` in `notWithAndNot` 
```

| `-` | Removes a specified element of any supported type from an array.
|===

> The following examples show uses of prepend, append, and remove operators on arrays:

Reviewer: `prepend`, `append`, and `remove` operators

[source,weave,linenums]
----
payload mapObject ((value,key) ->
    if(key as String == "age")
       {(key): value as Number + 1}
    else
       {(key): value} )
----

Using the update operator it is simplified to:

[source,weave,linenums]
----
payload update {
    case age at .age -> age + 1
}
----

As shown in the previous example with the update operator there is no need for casting or going through all the key value pairs. The update syntax is as follows:

[source,weave,linenums]
----
<value_to_update> update {
    case <variable_name> at <update_expression>[!]? [if(<conditional_expression>)]? -> <new value>
    ...
}
----

* `value_to_update` represents the original value that is going to be updated.
* `update_expression` is the selector expression that is going to match a given value of the `value_to_update`
* `variable_name` is the name of the variable that is going to bind to the matched `update_expression` existing value.
* `new_value` is the expression with the new value.
* `[if(<conditional_expression>)]?` If `conditional_expression` is true then the value is update. This section is optional and only required for Optional update.

> * `[!]` marks then selector as an upsert, which means that should the selector not be a match then it will be created.

Reviewer 3: My english is not very good, so I will rephrase this just to make sure the meaning is the same. 
When marking the expression with `!` if the expression doesn't match anything, the update operator is going to create all the required elements and insert with the new value. @kevintroller

Author: Updated.

* `update_expression` supports multiple `case` expressions.

[NOTE]
The update operator doesn't mutate the existing value but rather create a new one with the updated expressions.

=== Selector Expressions

> Use selector expressions with the update operator.

Reviewer 2: This sentence sounds weird to me, I think we should clarify to explain what selectors are used for, like `Use selector expressions with the update operator to specify which fields should be updated`

----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload update {
    case age at .age -> age + 1
    case s at .address.street -> "First Street"
}
----

.Output
[source,json,linenums]
----
{
  "name": "Ken",
  "lastName": "Shokida",
  "age": 31,
  "address": {
    "street": "First Street",
    "zipCode": "AB1234"
  }
}
----

==== Index Selector

This matching selector allows to match any array by its index. +

The following example updates an element that is at the first location of the array:

.Input
[source,json,linenums]
----
{
    "name": "Ken",
    "lastName":"Shokida",
    "age": 30,
    "addresses": [{
        "street": "Amenabar",
        "zipCode": "AB1234"
    }]
}
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload update {
    case s at .addresses[0] -> {
        "street": "Second Street",
        "zipCode": "ZZ123"
    }
}
----

.Output
[source,json,linenums]
----
{
  "name": "Ken",
  "lastName": "Shokida",
  "age": 30,
  "addresses": [
    {
      "street": "Second Street",
      "zipCode": "ZZ123"
    }
  ]
}
----

==== Attribute Selector

This matching selector allows to match any attribute value. +

> The following example updates an attribute value and change it to upper case:

Reviewer 2: `changes it`?

output application/json
---
payload map ((user) ->
    user update {
        case name at .name if(name == "Ken") -> name ++ " (Leandro)"
        case name at .name if(name == "Tomo") -> name ++ " (Christian)"
    }
)
----

.Output
[source,json,linenums]
----
[
  {
    "name": "Ken (Leandro)",
    "age": 30
  },
  {
    "name": "Tomo (Christian)",
    "age": 70
  },
  {
    "name": "Kajika",
    "age": 10
  }
]
----

=== Upserting

The update operator allows to insert a field if it is not present using the `!` symbol at the end of selector expression. When the field is not present the value `null` is going to be bound to the variable. +

The following example updates the category based on the name of the user:

.Input
[source,json,linenums]
----
[{"lastName": "Doe"}, {"lastName": "Parker", "name": "Peter" }]
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload  map ((user) ->
    user update {
        case name at .name! -> if(name == null) "JOHN" else upper(name)
    }
)
----

.Output
[source,json,linenums]
----
[
  {
    "lastName": "Doe",
    "name": "JOHN"
  },
  {
    "lastName": "Parker",
    "name": "PETER"
  }
]
----

=== Sugar Syntax

> The update operator allows to write a shorter version syntax to avoid some boilerplate code by default variable named `$`. +

Reviewer 2: `by using a` or `by defining a` or `by featuring a`

| `if else` a| An `if` operator evaluates a conditional expression and returns the value under the `if` only if the conditional expression is true. Otherwise, it returns the expression under `else`. Every `if` expression must have a matching `else` expression. See xref:dataweave-flow-control.adoc#control_flow_if_else[if else] for an example.
| `else if` | An `else` operator chains expressions together within an if-else construct by incorporating `else if`. See xref:dataweave-flow-control.adoc#control_flow_else_if[else if] for an example.
|===

== Update Operator

> DataWeave 2.0 supports the update operator that enables you to update a given field on a data structure with a new value. This use case is very common and was not simple to do in DataWeave.

Reviewer 4: ```suggestion
DataWeave supports the `update` operator, which enables you to update a given field of a data structure with a new value.
```

Reviewer 4: Also, the very first paragraph of the page should probably mention something about this functionality. "DataWeave 2.0 supports several mathematical, equality, relational, logical, prepend, append, **flow control, scope, and update** operators."

Reviewer 2: `a given field` -> should we say `specified fields` to make it clear several can be modified at once? @machaval

Author: Updated.

_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._ +

> Example: Using Update Operator, let's say that you have an object:

Reviewer 4: Suggest following the format of the examples in other sections on this page or picking a single format for them to follow.

Author: Done.

> Example: Using Update Operator, let's say that you have an object:

Reviewer 4: ```suggestion
Assume that you receive the following object as input :
```

Author: Updated.

[source,json,linenums]
----
{
    "name": "Ken",
    "lastName":"Shokida",
    "age": 30
}
----

> And you want to update the age by increasing it with one, without the need of recreating all the object again. Using the current DataWeave syntax is:

Reviewer 4: Note that `update` will be current in Mule 4.3. It sounds like we're doing a before/after comparison, but since we're working in the 4.3 branch, I think we just need to talk about the two ways you can accomplish the task in Mule 4.3/DW 2.3, with and without `update` as a way to help illustrate how to use the operator. 

The "And ..." is a sentence fragment. Can you perhaps turn that into a complete sentence? 

"To increase the age value by one without using the `update` operator or recreating the object again, you can use the following DataWeave syntax:"

Reviewer 2: I'd say perhaps we should mention something like `prior to 4.3, the way to accomplish this was...`

Author: Updated.

[source,weave,linenums]
----
payload mapObject ((value,key) ->
    if(key as String == "age")
       {(key): value as Number + 1}
    else
       {(key): value} )
----

> Using the update operator will be:

Reviewer 2: `will be` -> `it is simplified to`

Author: Updated.

[source,weave,linenums]
----
payload update {
    case age at .age -> age + 1
}
----

As shown in the previous example with the update operator there is no need for casting or going through all the key value pairs. The update syntax is as follows:

[source,weave,linenums]
----
<value_to_update> update {
    case <variable_name> at <update_expression>[!]? [if(<conditional_expression>)]? -> <new value>
    ...
}
----

* `value_to_update` represents the original value that is going to be updated.
* `update_expression` is the selector expression that is going to match a given value of the `value_to_update`
* `variable_name` is the name of the variable that is going to bind to the matched `update_expression` existing value.
* `new_value` is the expression with the new value.
* `[if(<conditional_expression>)]?` If `conditional_expression` is true then the value is update. This section is optional and only required for Optional update.

> * `[!]` marks then selector as an upsert, which means that if the selector is not matched then it is going to create it.

Reviewer 2: `which means that should the selector not be a match then it will be created`

Reviewer 2: or something like that, the current sentence sounds weird

Author: Updated.

* `update_expression` supports multiple `case` expressions.

[NOTE]
The update operator doesn't mutate the existing value but rather create a new one with the updated expressions.

=== Selector Expressions

Use selector expressions with the update operator.

==== Value Selector

This matching selector allows to match any field value of a given object. +

Example: Update a field at the root level and a field in a nested level using the value selector.

.Input
[source,json,linenums]
----
{
    "name": "Ken",
    "lastName":"Shokida",
    "age": 30,
    "address": {
        "street": "Amenabar",
        "zipCode": "AB1234"
    }
}
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload update {
    case age at .age -> age + 1

>     case s at .address.street -> "First Street"

Reviewer 2: @machaval should we make use of `s` in the example or maybe just not feature it?

.Input
[source,json,linenums]
----
{
    "name": "Ken",
    "lastName":"Shokida"
}
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
var theFieldName = "name"
---
payload update {
    case s at ."$(theFieldName)" -> "Shoki"
}
----

.Output
[source,json,linenums]
----
{
  "name": "Shoki",
  "lastName": "Shokida"
}
----

=== Conditional Update

The update operator has a syntax to write a conditional update if you only want to modify a field under a given condition. +

Example: Update the category based on the name of the user.

.Input
[source,json,linenums]
----
[{"name": "Ken", "age": 30}, {"name": "Tomo", "age": 70}, {"name": "Kajika", "age": 10}]
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload map ((user) ->
    user update {
        case name at .name if(name == "Ken") -> name ++ " (Leandro)"
        case name at .name if(name == "Tomo") -> name ++ " (Christian)"
    }
)
----

.Output
[source,json,linenums]
----
[
  {
    "name": "Ken (Leandro)",
    "age": 30
  },
  {
    "name": "Tomo (Christian)",
    "age": 70
  },
  {
    "name": "Kajika",
    "age": 10
  }
]
----

=== Upserting

> The update operator allows to insert a field if it is not present using the `!` symbol at the end of selector expression. When the field is not present the value `null` is going to be bind to the variable. +

Reviewer 2: `bound`?

Author: Updated.

Example: Update the category based on the name of the user.

.Input
[source,json,linenums]
----
[{"lastName": "Doe"}, {"lastName": "Parker", "name": "Peter" }]
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload  map ((user) ->
    user update {

>         case name at .name! -> if(name == null) "JHON" else upper(name)

Reviewer 2: `JOHN`

    }
)
----

.Output
[source,json,linenums]
----
[
  {
    "lastName": "Doe",

>     "name": "JHON"

Reviewer 2: same

  },
  {
    "lastName": "Parker",
    "name": "PETER"
  }
]
----

=== Sugar Syntax

> The update operator allows to write a shorter version syntax to avoid some code boilerplate and this is by binding a default variable name `$`. +

Reviewer 2: `boilerplate code`? `this is by` -> `by`, `named`

Author: Updated.

> Example: Same Value Selector example but using simple syntax.

Reviewer 2: `simpler`

| `if else` a| An `if` operator evaluates a conditional expression and returns the value under the `if` only if the conditional expression is true. Otherwise, it returns the expression under `else`. Every `if` expression must have a matching `else` expression. See xref:dataweave-flow-control.adoc#control_flow_if_else[if else] for an example.
| `else if` | An `else` operator chains expressions together within an if-else construct by incorporating `else if`. See xref:dataweave-flow-control.adoc#control_flow_else_if[else if] for an example.
|===

== Update Operator

> DataWeave supports the `update` operator, which enables you to update specified fields of a data structure with a new value.

Reviewer: ```suggestion
DataWeave 2.3.0 introduces and Mule 4.3 and later supports the `update` operator, which enables you to update specified fields of a data structure with new values.
```
fields = values (number agreement)

> _Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._ +

Reviewer: Suggest moving this to the narrative above, as shown, to avoid the weird italic pseudo-heading.

> The following example shows use of the update operator, assume that you receive the next object as input:

Reviewer: ```suggestion
For the following example showing use of the `update` operator, assume that you receive the following object as input:
```

[source,json,linenums]
----
{
    "name": "Ken",
    "lastName":"Shokida",
    "age": 30
}
----

> Prior to Mule 4.3, to increase the age value by one without using the update operator or recreating the object again, the way to accomplish this was using the following DataWeave syntax:

Reviewer: ```suggestion
Prior to Mule 4.3, you would have to use the following DataWeave syntax to increase the `age` value by 1 without using the `update` operator or re-creating the object:
```

[source,weave,linenums]
----
payload mapObject ((value,key) ->
    if(key as String == "age")
       {(key): value as Number + 1}
    else
       {(key): value} )
----

> Using the update operator it is simplified to:

Reviewer: ```suggestion
Using the `update` operator simplifies the syntax to accomplish the same task:
```

[source,weave,linenums]
----
payload update {
    case age at .age -> age + 1
}
----

> As shown in the previous example with the update operator there is no need for casting or going through all the key value pairs. The update syntax is as follows:

Reviewer: I think what you want to say is this. The wording might need tweaking to meet your intentions better.

"Using the `update` operator also simplifies the process of updating multiple values:"

[source,weave,linenums]
----
<value_to_update> update {
    case <variable_name> at <update_expression>[!]? [if(<conditional_expression>)]? -> <new value>
    ...
}
----

> * `value_to_update` represents the original value that is going to be updated.

Reviewer: ```suggestion
* `value_to_update` is the current value, which you want to update.
```

> * `update_expression` is the selector expression that is going to match a given value of the `value_to_update`

Reviewer: ```suggestion
* `update_expression` is a selector expression that matches the value of `value_to_update`.
```

> * `variable_name` is the name of the variable that is going to bind to the matched `update_expression` existing value.

Reviewer: ```suggestion
* `variable_name` identifies the variable to bind to the matched `update_expression` value.
```

> * `new_value` is the expression with the new value.

Reviewer: ```suggestion
* `new_value` is the value you want to replace `value_to_update`.
```

> * `[if(<conditional_expression>)]?` If `conditional_expression` is true then the value is update. This section is optional and only required for Optional update.

Reviewer: ```suggestion
* `[if(<conditional_expression>)]?` is required only for optional updates that occur only if the value of `conditional_expression` is true.
```

> * `[!]` marks the selector as an upsert. When marking the expression with `!` if the expression doesn't match anything, the update operator creates all the required elements and inserts the new value.

Reviewer: ```suggestion
* `[!]` marks the selector as an upsert, which either updates the expression, if it exists, or creates all the required elements and inserts a new expression if not.
```

* `update_expression` supports multiple `case` expressions.

[NOTE]

> The update operator doesn't mutate the existing value but rather create a new one with the updated expressions.

Reviewer: ```suggestion
The `update` operator doesn't mutate the existing value but rather creates a new one with the updated expressions.
```

=== Selector Expressions

> Use selector expressions with the update operator to specify which fields should be updated.

Reviewer: ```suggestion
Use selector expressions with the `update` operator to specify which fields should be updated.
```

==== Value Selector

> This matching selector allows to match any field value of a given object. +

Reviewer: ```suggestion
The value selector enables you to match the value of any field of a given object. +
```
Should this be `value` selector?

> The following example updates a field at the root level and a field in a nested level using the value selector:

Reviewer: Is there supposed to be something that says "value" in the example? I was looking for it, but maybe I don't understand matching selectors.

.Input
[source,json,linenums]
----
{
    "name": "Ken",
    "lastName":"Shokida",
    "age": 30,
    "address": {
        "street": "Amenabar",
        "zipCode": "AB1234"
    }
}
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload update {
    case age at .age -> age + 1
    case s at .address.street -> "First Street"
}
----

.Output
[source,json,linenums]
----
{
  "name": "Ken",
  "lastName": "Shokida",
  "age": 31,
  "address": {
    "street": "First Street",
    "zipCode": "AB1234"
  }
}
----

==== Index Selector

> This matching selector allows to match any array by its index. +

Reviewer: ```suggestion
The index selector enables you to match any array by its index. +
```

> The following example updates an element that is at the first location of the array:

Reviewer: ```suggestion
The following example updates the first element of an array:
```

.Input
[source,json,linenums]
----
{
    "name": "Ken",
    "lastName":"Shokida",
    "age": 30,
    "addresses": [{
        "street": "Amenabar",
        "zipCode": "AB1234"
    }]
}
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload update {
    case s at .addresses[0] -> {
        "street": "Second Street",
        "zipCode": "ZZ123"
    }
}
----

.Output
[source,json,linenums]
----
{
  "name": "Ken",
  "lastName": "Shokida",
  "age": 30,
  "addresses": [
    {
      "street": "Second Street",
      "zipCode": "ZZ123"
    }
  ]
}
----

==== Attribute Selector

> This matching selector allows to match any attribute value. +

Reviewer: ditto prev edits to intro sentence

> The following example updates an attribute value and changes it to upper case:

Reviewer: ```suggestion
The following example updates an attribute value to be presented in uppercase letters:
```

.Input
[source,xml,linenums]
----
<user name="leandro"/>
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload update {
    case name at .user.@name -> upper(name)
}
----

.Output
[source,xml,linenums]
----
<?xml version='1.0' encoding='UTF-8'?>
<user name="LEANDRO"/>
----

==== Dynamic Selector

> Any of the given selectors can use an expression to make it dynamic. +

Reviewer: ```suggestion
You can make any selector dynamic. +
```
Do we need to explain what we mean by "dynamic"?

> The following example selects and updates a value where the field to be updated is in a variable and not hardcoded. In this case from the variable `theFieldName` with value `name`:

Reviewer: ```suggestion
The following example selects and updates a field that is in a variable (`theFieldName`), which is not hardcoded, to the value `name`:
```

.Input
[source,json,linenums]
----
{
    "name": "Ken",
    "lastName":"Shokida"
}
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
var theFieldName = "name"
---
payload update {
    case s at ."$(theFieldName)" -> "Shoki"
}
----

.Output
[source,json,linenums]
----
{
  "name": "Shoki",
  "lastName": "Shokida"
}
----

=== Conditional Update

> The update operator has a syntax to write a conditional update if you only want to modify a field under a given condition. +

Reviewer: ```suggestion
You can configure the `update` operator to perform a conditional update if you want to modify a field only under a given condition. +
```

The following example updates the category based on the name of the user:

.Input
[source,json,linenums]
----
[{"name": "Ken", "age": 30}, {"name": "Tomo", "age": 70}, {"name": "Kajika", "age": 10}]
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload map ((user) ->
    user update {
        case name at .name if(name == "Ken") -> name ++ " (Leandro)"
        case name at .name if(name == "Tomo") -> name ++ " (Christian)"
    }
)
----

.Output
[source,json,linenums]
----
[
  {
    "name": "Ken (Leandro)",
    "age": 30
  },
  {
    "name": "Tomo (Christian)",
    "age": 70
  },
  {
    "name": "Kajika",
    "age": 10
  }
]
----

=== Upserting

> The update operator allows to insert a field if it is not present using the `!` symbol at the end of selector expression. When the field is not present the value `null` is going to be bound to the variable. +

Reviewer: ```suggestion
The `update` operator enables you to insert a field if it is not present by using the `!` symbol at the end of the selector expression.+
```
The current second sentence seems to be an (unnecessary?) explanation of WHY you want to do this, which has not appeared for any of the other sections.
If we need to give a reason, I suggest rewording it to something like "Inserting a missing field avoids binding the `null` value to the variable."   [Which variable?]

> The following example updates the category based on the name of the user:

Reviewer: So...I don't understand how this sentence relates to upserting, and I don't see the `!` symbol.
I expect the example to illustrate the conceptual info above it.

.Input
[source,json,linenums]
----
[{"lastName": "Doe"}, {"lastName": "Parker", "name": "Peter" }]
----

.Source
[source,weave,linenums]
----
%dw 2.0
output application/json
---
payload  map ((user) ->
    user update {
        case name at .name! -> if(name == null) "JOHN" else upper(name)
    }
)
----

.Output
[source,json,linenums]
----
[
  {
    "lastName": "Doe",
    "name": "JOHN"
  },
  {
    "lastName": "Parker",
    "name": "PETER"
  }
]
----

> 

Reviewer: P.S. I am TOTALLY OKAY with using Peter Parker as a general user name. I just think it's risky to rely on the relationships implied in the example that used three Avengers characters. Sigh. Sometimes I think I should just edit everyone individually and not try to send reminders about rules that can be broken based on context....

=== Sugar Syntax

> The update operator allows to write a shorter version syntax to avoid some boilerplate code by using a default variable named `$`. +

Reviewer: ```suggestion
The `update` operator enables you to shorten version syntax by using a default variable named `$` to avoid boilerplate code. +
```

> The following example uses the same Value selector example but using simpler syntax:

Reviewer: ```suggestion
The following example illustrates how to simplify the syntax used for value selector:
```
I was not sure which example provided earlier was "the same Value selector example." As always, tweak suggested wordings to maintain tech accuracy.
